internal/certs/certs.go

```diff
@@ -0,0 +1,156 @@
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+
+	newCertificatePrivateKey, err := rsa.GenerateKey(rand.Reader, RsaKeyBits())
+	if err != nil {
+		return []byte{}, nil, err
+	}
+
+	newCertificate, err := x509.CreateCertificate(rand.Reader, newCertificateTemplate, newCertificateTemplate, &newCertificatePrivateKey.PublicKey, newCertificatePrivateKey)
+	if err != nil {
+		return []byte{}, nil, err
+	}
+
+	newCertificatePem := new(bytes.Buffer)
+	pem.Encode(newCertificatePem, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: newCertificate,
+	})
+
+	newCertificatePrivateKeyPem := new(bytes.Buffer)
+	pem.Encode(newCertificatePrivateKeyPem, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(newCertificatePrivateKey),
+	})
+
+	return newCertificatePem.Bytes(), newCertificatePrivateKeyPem.Bytes(), nil
+}
+
+func GenerateSignedCertificate(expiration time.Time, sourceCertificate, sourcePrivateKey []byte) ([]byte, []byte, error) {
+	newCertificateTemplate := &x509.Certificate{
+		SerialNumber: getRandomInt(),
+		DNSNames:     getDns(),
+		NotBefore:    time.Now(),
+		NotAfter:     expiration,
+		ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:     x509.KeyUsageDigitalSignature,
+	}
+
+	newCertificatePrivateKey, err := rsa.GenerateKey(rand.Reader, RsaKeyBits())
+	if err != nil {
+		return []byte{}, nil, err
+	}
+
+	decodedSourceCertificate, _ := pem.Decode(sourceCertificate)
+	parsedSourceCertificate, err := x509.ParseCertificate(decodedSourceCertificate.Bytes)
+	if err != nil {
+		return []byte{}, nil, err
+	}
+	decodedSourcePrivateKey, _ := pem.Decode(sourcePrivateKey)
+	parsedSourcePrivateKey, err := x509.ParsePKCS1PrivateKey(decodedSourcePrivateKey.Bytes)
+	if err != nil {
+		return []byte{}, nil, err
+	}
+
+	newCertificate, err := x509.CreateCertificate(rand.Reader, newCertificateTemplate, parsedSourceCertificate, &newCertificatePrivateKey.PublicKey, parsedSourcePrivateKey)
+	if err != nil {
+		return []byte{}, nil, err
+	}
+
+	newCertificatePem := new(bytes.Buffer)
+	pem.Encode(newCertificatePem, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: newCertificate,
+	})
+
+	newCertificatePrivateKeyPem := new(bytes.Buffer)
+	pem.Encode(newCertificatePrivateKeyPem, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(newCertificatePrivateKey),
+	})
+
+	return newCertificatePem.Bytes(), newCertificatePrivateKeyPem.Bytes(), nil
+}
+
+func VerifyIfSecondCertificateIsSignedByFirstCertificate(firstCertificate, secondCertificate []byte) (bool, error) {
+	firstCertificateDecoded, _ := pem.Decode(firstCertificate)
+	secondCertificateDecoded, _ := pem.Decode(secondCertificate)
+
+	firstCertificateTemplate, err := x509.ParseCertificate(firstCertificateDecoded.Bytes)
+	if err != nil {
+		return false, err
+	}
+	if !firstCertificateTemplate.IsCA {
```

Comment:
Should we check also if it is self-signed CA root? Or just checking if it is CA cert is enough?